File: _experimental/kotlin/OAuthRequest.kt
// Kotlin Vanilla OAuth Client Example for UserSpice
// Simulates initiating the OAuth request

import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import java.util.UUID

fun main() {
    val settings = OAuthClientConfig.getSettings()

    val authorizationEndpoint = settings["server_url"] + settings["auth_endpoint_path"]

    // Generate a random state parameter for CSRF protection
    val state = UUID.randomUUID().toString()
    // In a real web app, you would store this 'state' in the user's session
    // to verify it on callback.
    println("Generated state (should be stored in session): $state")

    try {
        // Build the authorization URL
        val authUrl = buildString {
            append(authorizationEndpoint)
            append("?response_type=code")
            append("&client_id=${URLEncoder.encode(settings["client_id"], StandardCharsets.UTF_8.name())}")
            append("&redirect_uri=${URLEncoder.encode(settings["redirect_uri"], StandardCharsets.UTF_8.name())}")
            append("&state=${URLEncoder.encode(state, StandardCharsets.UTF_8.name())}")
            append("&scope=${URLEncoder.encode("profile", StandardCharsets.UTF_8.name())}") // Add any scopes you need
        }

        println("\n--- OAuth Authorization Request Simulation (Kotlin) ---")
        println("In a real web application, you would redirect the user to the following URL:")
        println(authUrl)
        println("\nAfter the user authorizes, they will be redirected to your redirect_uri with a 'code' and 'state' parameter.")
        println("Example: ${settings["redirect_uri"]}?code=AUTHORIZATION_CODE_HERE&state=$state")

    } catch (e: Exception) {
        System.err.println("Error building authorization URL: ${e.message}")
        e.printStackTrace()
    }
}

File: _experimental/perl/oauth_request.pl
#!/usr/bin/perl
# Perl Vanilla OAuth Client Example for UserSpice
# Simulates initiating the OAuth request

use strict;
use warnings;
use LWP::UserAgent; # For making HTTP requests (though not used directly here, good practice for response)
use URI;
use URI::Escape;
use Digest::MD5 qw(md5_hex); # Simple way to get a somewhat random string for state

# Assuming oauth_config.pl is in the same directory or in @INC
require "oauth_config.pl";
my $settings = get_oauth_settings();

my $authorization_endpoint = $settings->{server_url} . $settings->{auth_endpoint_path};

# Generate a state parameter for CSRF protection
# Using time and a random number, then hashing for simplicity. UUID module would be better.
my $state = md5_hex(time() . rand() . $$);
# In a real web app, you would store this 'state' in the user's session
# to verify it on callback.
print "Generated state (should be stored in session): $state\n";

# Build the authorization URL
my $auth_uri = URI->new($authorization_endpoint);
$auth_uri->query_form(
    response_type => 'code',
    client_id     => $settings->{client_id},
    redirect_uri  => $settings->{redirect_uri},
    state         => $state,
    scope         => 'profile', # Add any scopes you need
);

my $auth_url = $auth_uri->as_string;

print "\n--- OAuth Authorization Request Simulation (Perl) ---\n";
print "In a real web application, you would redirect the user to the following URL:\n";
print "$auth_url\n";
print "\nAfter the user authorizes, they will be redirected to your redirect_uri with a 'code' and 'state' parameter.\n";
print "Example: $settings->{redirect_uri}?code=AUTHORIZATION_CODE_HERE&state=$state\n";

exit 0;
